src/redisearch_rs/rqe_iterators/src/optional.rs

```diff
@@ -0,0 +1,174 @@
+        Self {
+            max_doc_id: max_id,
+            weight,
+
+            last_doc_id: 0,
+            result: RSIndexResult::virt().frequency(1),
+
+            child: Some(child),
+        }
+    }
+}
+
+impl<'index, I> RQEIterator<'index> for Optional<'index, I>
+where
+    I: RQEIterator<'index>,
+{
+    fn read(&mut self) -> Result<Option<&mut RSIndexResult<'index>>, RQEIteratorError> {
+        if self.at_eof() {
+            return Ok(None);
+        }
+
+        // The second condition of this if statement is
+        // to catch edge cases where the child is only _just_
+        // now exhausted while it wasn't before and thus returned
+        // Nothing. This way we still fall back to Virtual results
+        // in this case instead of returning None incorrectly.
+        //
+        // C Version only had the last_doc_id equality check
+        // but that's because it has access to the result of child
+        // even in cases like the one described here, while the
+        // Rust version only exposes the result if the
+        // ierator is not EOF.
+        //
+        // The C version also used to do:
+        //
+        // ```
+        // self.last_doc_id == self.child.last_doc_id()
+        // ```
+        //
+        // Here we instead assume the child handles this fine,
+        // as why wouldn't it?
+        if let Some(outcome) = self.child.read()? {
+            self.last_doc_id = outcome.doc_id;
+            outcome.weight = self.weight;
+            return Ok(Some(outcome));
+        } else {
+            self.last_doc_id += 1;
+        }
+
+        self.result.doc_id = self.last_doc_id;
+        Ok(Some(&mut self.result))
+    }
+
+    // [OG C Comment] SkipTo for OPTIONAL iterator - Non-optimized version.
+    // Skip to a specific docId. If the child has a hit on this docId, return it.
+    // Otherwise, return a virtual hit.
+    //
+    // TODO: Optimized version will also be integrated here, but in a later PR.
+    fn skip_to(
+        &mut self,
+        doc_id: t_docId,
+    ) -> Result<Option<SkipToOutcome<'_, 'index>>, RQEIteratorError> {
+        debug_assert!(doc_id > self.last_doc_id);
+
+        if doc_id > self.max_doc_id || self.at_eof() {
+            self.last_doc_id = doc_id;
+            return Ok(None);
+        }
+        if doc_id > self.child.last_doc_id()
+            && let Some(SkipToOutcome::Found(ris)) = self.child.skip_to(doc_id)?
+            && ris.doc_id == doc_id
+        {
+            // real hit
+            self.last_doc_id = ris.doc_id;
+            ris.weight = self.weight;
+            return Ok(Some(SkipToOutcome::Found(ris)));
+        }
+
+        // virtual hit
+        self.last_doc_id = doc_id;
+        self.result.doc_id = self.last_doc_id;
+        Ok(Some(SkipToOutcome::Found(&mut self.result)))
+    }
+
+    fn revalidate(&mut self) -> RQEValidateStatus {
+        match self.child.revalidate() {
+            RQEValidateStatus::Ok => RQEValidateStatus::Ok,
```

Comment:
The logic here is different from the C code where we return `Ok` only if the current result is virtual. Shouldn't we check that as well?

```c
  if (base->current == oi->virt || child_status == VALIDATE_OK) {
```

Reply:
Correct... But we do not work with comparing results in Rust.
So not sure what would be the equivalent logic, any concrete suggestion?

Reply:
can't we just do this?

```rust
if self.result.kind() == RSResultKind::Virtual {}
```

Reply:
`self.result` is always virtual... As we do not assign the result from child one like we do in C.